### ISSUE TITLE
"ejabberdctl join_cluster" causes problems

### ISSUE THREAD

**Issue by dakom** [processone/ejabberd#676]
Following the docs at http://docs.ejabberd.im/admin/guide/clustering/ I do the following:

`first node`
1. sudo ejabberdctl start

`second node`
1. sudo ejabberdctl start
2. sudo ejabberdctl join_cluster 'ejabberd@first'

and... I get a crash. However, if I restart the server (removing init scripts) and run the join_cluster command *before*  start (in fact before *any* ejabberdctl command)- then everything is fine (and at a glance it seems messages go to both nodes- though I haven't tested in-depth)

However, it's not as simple as simply reversing those instructions in the docs or killing epmd or something, because in that case it seems the configuration is not saved. Specifically, If I put the init script back, i.e. `sudo update-rc.d -f ejabberd defaults` (using the ejabberd.init from the build folder that was copied to /etc/init.d/ejabberd) - then restart the server, it starts without errors in the log but **messages do not seem to go to both nodes** (however the website still shows the second node in the panel)

Using latest build and erlang v18

**Comment by dakom** [comment]
When following the original suggestion (join_cluster *after* ejabberd is already started), the crash error begins:

>Protocol: ~tp: the name ejabberd@second seems to be in use by another Erlang node",["inet_tcp"]}

**Comment by dakom** [comment]
Hmmm... after implementing the fix from #677 I am now seeing different types of errors, probably the good kind since they reference the other node and mnesia :)

Going to wipe clean and try again... I might have just corrupted something along the way due to a misconfigured install

**Comment by dakom** [comment]
OK, with a fresh install following that configuration change- I *still* get the same bug described at the top here (starting ejabberd on 2 different fresh vms that have the same `/root/.erlang-cookie`, then attempting a `sudo ejabberdctl join_cluster 'ejabberd@first'`, from one to the other, causes a crash)

However- if I do not start ejabberd initially, and rather just reboot and run the `join_cluster` command right away- I get a different error:

> node ejabberd@first is not reachable, please check epmd port, and FIREWALL_WINDOW ports

This is strange, since `netcat -v first 4369` gives me:
> Connection to first 4369 port [tcp/epmd] succeeded!

(fwiw in my actual setup it's not called "first", in case that's some special reserved name I'm supposed to set my hostname to or something... I highly doubt that but mentioning it just in case)

**Comment by dakom** [comment]
@cromain  - have you been able to duplicate the error? (just wondering in case it's just something I missed or if it's an actual bug). Thanks!

**Comment by mremond** [comment]
We will have a look but `join_cluster` is working. We are using it in production. We will see if there is need for clarifications in the documentation.

**Comment by dakom** [comment]
OK cool, maybe I just happened check out a snapshot that was buggy or there's some issue with using google compute engine? I might not be able to get to it for a day or two, but when I do I'll try from scratch and log each step so you can duplicate the process/bug exactly (or spot where I went wrong). Thanks again!

**Comment by cromain** [comment]
ejabberdctl must run as user, not root. if running as root, then ejabberdctl does this "su $INSTALLUSER -c ...", so this is not a good practice to call "sudo ejabberdctl ..."

calling ejabberdctl join_cluster is just a wrapper to call bin/joincluster script. (i just fixed a PATH issue in ejabberdctl packaged in our installer (will be available in 15.09), but not sure your using this one anyway)

so procedure is:
log as ejabberd user on first, and ejabberdctl start (ejabberd@first is running)
log as ejabberd user on second, and call 'ejabberdctl join_cluster ejabberd@first' or 'bin/joincluster ejabberd@first'
then ejabberdctl start (ejabberd@second is running)
on debug, if you type mnesia:info(). you can see cluster is ok:
running db nodes   = [ejabberd@first,ejabberd@second]

the joincluster script is working as expected.

**Comment by joudinet** [comment]
Hi,

I also get the following error message if I run ejabberdctl join_cluster as root:
> node ejabberd@first is not reachable, please check epmd port, and FIREWALL_WINDOW ports

@cromain indication to run it as ejabberd user instead of root did work, but I believe you should improve the error message, or even find a way to make it works when executed with "su $INSTALLUSER -c"

**Comment by shobanapeter** [comment]
The join_cluster command intermittently fails to start the ejabberd process which leaves the node in "NOT RUNNING" state.
su ejabberd -c "/opt/ejabberd/bin/ejabberdctl join_cluster ejabberd@$ejabberd-xmpp-1"
I executed the above command and ejabberd process hung.
[root@ejabberd-xmpp-0 logs]# service ejabberd status
Getting ejabberd status...
**The node 'ejabberd@ejabberd-xmpp-0' is started with status: started
ejabberd is not running in that node**
Check for error messages: /opt/ejabberd/logs/ejabberd.log
or other files in that directory.

We've worked around the problem by restarting the ejabberd process after join_cluster. Is this a bug in join_cluster? or is it a mandate to call ejabberd restart after join_cluster?
